gax/src/main/java/com/google/api/gax/rpc/ClientContext.java

```diff
@@ -134,26 +134,34 @@ public static ClientContext create(ClientSettings settings) throws IOException {
    * settings.
    */
   public static ClientContext create(StubSettings settings) throws IOException {
+    String quotaProjectID = null;
     ApiClock clock = settings.getClock();
 
     ExecutorProvider executorProvider = settings.getExecutorProvider();
     final ScheduledExecutorService executor = executorProvider.getExecutor();
 
+    Map<String, String> headers = getHeadersFromSettings(settings);
+    if (headers.containsKey(QUOTA_PROJECT_ID_HEADER_KEY)) {
+      quotaProjectID = headers.get(QUOTA_PROJECT_ID_HEADER_KEY);
+    }
+
     Credentials credentials = settings.getCredentialsProvider().getCredentials();
 
-    if (settings.getQuotaProjectId() != null) {
+    if (settings.getQuotaProjectId() != null || quotaProjectID != null) {
       // If the quotaProjectId is set, wrap original credentials with correct quotaProjectId as
       // QuotaProjectIdHidingCredentials.
       // Ensure that a custom set quota project id takes priority over one detected by credentials.
       // Avoid the backend receiving possibly conflict values of quotaProjectId
       credentials = new QuotaProjectIdHidingCredentials(credentials);
+    } else if (credentials != null
+        && credentials.getRequestMetadata().containsKey(QUOTA_PROJECT_ID_HEADER_KEY)) {
+      quotaProjectID = credentials.getRequestMetadata().get(QUOTA_PROJECT_ID_HEADER_KEY).get(0);
```

Comment:
I don't think this else clause is needed for the reason described in https://github.com/googleapis/gax-java/pull/1166/files#diff-ac347ed639ae3cbdd394c7769616ff1dR224

Reply:
credentials could null when NoCredentialProvider is used in the unit test.

Reply:
For this one I mean updating the quotaProjectId value to be set on the ClientContext.Builder at the bottom of ClientContext.create.